pyomo/contrib/mindtpy/cut_generation.py

```diff
@@ -181,6 +181,53 @@ def add_oa_cuts(
                         )
 
 
+def add_oa_cuts_for_grey_box(
+    target_model, jacobians_model, config, objective_sense, mip_iter, cb_opt=None
+):
+    sign_adjust = -1 if objective_sense == minimize else 1
+    if config.add_slack:
+        slack_var = target_model.MindtPy_utils.cuts.slack_vars.add()
+    for target_model_grey_box, jacobian_model_grey_box in zip(
+        target_model.MindtPy_utils.grey_box_list,
+        jacobians_model.MindtPy_utils.grey_box_list,
+    ):
+        jacobian_matrix = (
+            jacobian_model_grey_box.get_external_model()
+            .evaluate_jacobian_outputs()
+            .toarray()
+        )
+        for index, output in enumerate(target_model_grey_box.outputs.values()):
+            dual_value = jacobians_model.dual[jacobian_model_grey_box][
+                output.name.replace("outputs", "output_constraints")
+            ]
+            target_model.MindtPy_utils.cuts.oa_cuts.add(
+                expr=copysign(1, sign_adjust * dual_value)
+                * (
+                    sum(
+                        jacobian_matrix[index][var_index] * (var - value(var))
+                        for var_index, var in enumerate(
+                            target_model_grey_box.inputs.values()
+                        )
+                    )
+                )
+                - (output - value(output))
+                - (slack_var if config.add_slack else 0)
+                <= 0
+            )
+    # TODO: gurobi_persistent currently does not support greybox model.
```

Comment:
Should this become an issue?

Reply:
`Gurobi` will fail even when we deactivate the `greybox` block. Therefore, we cannot use `Gurobi` as the mip solver.

Reply:
I will submit an issue related to this.

Reply:
https://github.com/Pyomo/pyomo/issues

Reply:
@emma58 Could you provide me a `reference` example?

Reply:
```
from pyomo.environ import *

m = ConcreteModel()
m.b = Block()
m.b.x = Var(bounds=(2, 5))
m.y = Var(domain=NonNegativeReals)
m.c = Constraint(expr=m.b.x - m.y <= 4)
m.obj = Objective(expr=m.b.x + m.y)

m.b.deactivate()

m.somewhere_safe_to_keep_things = Block()
m.somewhere_safe_to_keep_things.x = Reference(m.b.x)

opt = SolverFactory('gurobi_persistent')
opt.set_instance(m)

opt.solve(m, tee=True)
```
This is the main idea. You'll want to figure out how to do your own accounting in terms of finding the Vars you need to reference and naming the references, etc.

Reply:
#3000 Issue opened

Reply:
@emma58 . I don't think `Reference` is an elegant way to resolve this issue. I will comment the code of adding `Gurobi` lazy constraints and only support `greybox` in multi-tree implementation this time. 
Btw, since persistent solver will be totally replaced by `appsi_solver`, does appsi_solver support both `CPLEX` and `Gurobi` callbacks as `persistent solver` now? @michaelbynum

Reply:
appsi_gurobi does support callbacks. appsi_cplex will, but does not currently.